asllib/doc/ASLSemanticsReference.tex

```diff
@@ -19,160 +266,641 @@
 
-The semantics of ASL are embodied in a function that takes as input an ASL program
-and returns all possible behaviours of that program.
+The semantics of ASL defines all valid behaviors of a given ASL specification.
+More precisely, an ASL specification is first parsed into an \emph{abstract syntax tree},
+or AST, for short. Second, a type checker analyzes the \emph{parsed AST} for correctness and, if successful,
+returns a \emph{static environment} and a \emph{typed AST}. Finally, tools such as interpreters and verifiers
+can operate over the typed AST, based on the definition of the semantics, to test and analyze a given specification.
+
+The abstract syntax, parsed AST, and typed AST are defined in the abstract syntax reference~\cite{ASLAbstractSyntaxReference}.
+The ASL type system, which defines the set of parsed ASTs that are considered valid and how to
+produce the corresponding typed ASTs and static environments is defined in the typing reference~\cite{ASLTypingReference}.
+
+\paragraph{Basic Semantic Concepts.}
+The semantics is constructively defined via \emph{configurations}, \emph{transitions}, and \emph{semantic rules}.
+Intuitively, configurations represent intermediate states during the execution of a program and they encapsulate
+the information needed to transition into other configurations.
+Configurations include a \emph{dynamic environment}, which binds variables to values;
+the typed AST node that needs to be ``evaluated''; a \emph{mode}, which distinguishes between different kinds of states, for example
+states where an exception was raised and states where no exception was raised; a \herd\ \emph{execution graph} in the case of a
+\emph{concurrent semantics}; and possible other bits of information, for example intermediate expression values.
+Semantic rules define the set of possible transitions.
+
+\paragraph{Execution.}
+A valid execution of an ASL specification transitions from an \emph{initial configuration} to an output configuration.
+Such a transition is defined by transitions over the AST nodes making up the specification.
+
+\paragraph{Primitive Subprograms.}
+The semantics of ASL is parameterized by a set of primitive subprograms ---
+subprograms whose implementation is not defined by an ASL body and whose effect on the dynamic environment
+is defined externally. Critically, access to memory is given by primitive subprograms.
+
+We define two types of semantics --- a \emph{sequential semantics} and a \emph{concurrent semantics}.
+Each semantics operates over a different kind of environment.
+
+\paragraph{Sequential Semantics.}
+The sequential semantics corresponds to executing an ASL specification in the context of a single (conceptual) process
+of execution --- notice that ASL does not contain any concurrency constructs. Access to memory is hidden away
+in primitive subprograms that interface with a map-like memory implementation that does not reorder memory accesses
+and serves one access at a time.
```

Comment:
How does sequential aslref behave when calling the primitive memory access subprograms?

Reply:
I think this could be resolved by simply dropping the last sentence:
"Access to memory is hidden away in primitive subprograms that interface with a map-like memory implementation that does not reorder memory accesses and serves one access at a time."